### PR TITLE
Cmake: Add 3.28:

### PR DESCRIPTION
**Pull request by johnwparent**

### CODE REVIEW

**Comment by johnwparent** [comment]
cc @alalazo

**Review by tldahlgren — COMMENTED** [review]
FYI. Confirmed the 5 new version sha256s.

**Comment by cedricchevalier19** [comment]
@johnwparent can you describe a little bit more the conflict with Kokkos?

We have observed an issue using CMake 3.28.4 and Cuda (https://github.com/kokkos/kokkos/pull/6898), but it should be solved in the last version 4.3.00 of Kokkos.

Thank you very much in advance.

**Comment by johnwparent** [comment]
> FYI. Confirmed the 5 new version sha256s.

Thanks as always!!

**Comment by johnwparent** [comment]
> @johnwparent can you describe a little bit more the conflict with Kokkos?
> 
> We have observed an issue using CMake 3.28.4 and Cuda ([kokkos/kokkos#6898](https://github.com/kokkos/kokkos/pull/6898)), but it should be solved in the last version 4.3.00 of Kokkos.
> 
> Thank you very much in advance.

@cedricchevalier19 

With CMake 3.28 (all patch versions so far) we're seeing a CI failure specifically with `kokkos-kernels` and `trilinos` right now, but this would be true with any package consuming the `kokkos` CMake config file. The error we observe is an eventual make (or CMake for 3.28.4) failure caused by this (from the kokkos config):  

```
  CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0111 is not set: An imported target missing its location property
  fails during generation.  Run "cmake --help-policy CMP0111" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  IMPORTED_LOCATION not set for imported target "CUDA::cuda_driver"
  configuration "RELEASE".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at kokkos/lib64/cmake/Kokkos/KokkosConfig.cmake:41 (SET_TARGET_PROPERTIES):
  The link interface of target "CUDA::cudart" contains:

    CUDA::cudart_static_deps

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

  Call Stack (most recent call first):
  CMakeLists.txt:129 (FIND_PACKAGE)
```
when tools eventually can't resolve the path to that target.
From a grep of the kokkos code base we can see some custom logic around this and similar targets
```
grep -rin "cudart_static" .
./cmake/Modules/CudaToolkit.cmake:130:- ``CUDA::cudart_static``
./cmake/Modules/CudaToolkit.cmake:806:  _CUDAToolkit_find_and_add_import_lib(cudart_static)
./cmake/Modules/CudaToolkit.cmake:808:  # setup dependencies that are required for cudart_static when building
./cmake/Modules/CudaToolkit.cmake:811:  if(NOT TARGET CUDA::cudart_static_deps
./cmake/Modules/CudaToolkit.cmake:812:     AND TARGET CUDA::cudart_static)
./cmake/Modules/CudaToolkit.cmake:814:    add_library(CUDA::cudart_static_deps IMPORTED INTERFACE)
./cmake/Modules/CudaToolkit.cmake:815:    target_link_libraries(CUDA::cudart_static INTERFACE CUDA::cudart_static_deps)
./cmake/Modules/CudaToolkit.cmake:819:      target_link_libraries(CUDA::cudart_static_deps INTERFACE Threads::Threads ${CMAKE_DL_LIBS})
./cmake/Modules/CudaToolkit.cmake:827:        message(WARNING "Could not find librt library, needed by CUDA::cudart_static")
./cmake/Modules/CudaToolkit.cmake:829:        target_link_libraries(CUDA::cudart_static_deps INTERFACE ${CUDAToolkit_rt_LIBRARY})
```
So it seems like when we updated the CUDA detection behavior in CMake, the Kokkos logic around these targets broke and is now producing bad KokkosConfig.cmake files


@cedricchevalier19 is the Fix in Kokkos in a release that's part of Spack? If so I'll scope the conflict appropriately.

**Comment by cedricchevalier19** [comment]
Fix is present in `4.3.00` which is available in Spack.

For other versions, `conflict` should be declared only when `+cuda` is set.

**Comment by johnwparent** [comment]
> Fix is present in `4.3.00` which is available in Spack.
> 
> For other versions, `conflict` should be declared only when `+cuda` is set.

@cedricchevalier19 
Thank you, I've updated the conflict to be more specific.

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 1]
```diff
@@ -50,6 +50,7 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
     version("3.0.00", sha256="c00613d0194a4fbd0726719bbed8b0404ed06275f310189b3493f5739042a92b")
 
     depends_on("cmake@3.16:", type="build")
+    conflicts("cmake@3.28:", when="@:4.2.01 +cuda")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/kokkos/package.py]: I'll submit this separately, since it might be needed also in #43349

**Review by cedricchevalier19 — APPROVED** [review]
It is fine on the Kokkos side. I'll let @alalazo and the spack team decide if they want to separate the Kokkos `conflicts` modification.

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 1 above)
Comment (johnwparent) [reply var/spack/repos/builtin/packages/kokkos/package.py]: From the GL CI, this change is not required for CMake 3.29
https://gitlab.spack.io/spack/spack/-/jobs/10789546

**Comment by johnwparent** [comment]
Note: requires https://github.com/spack/spack/pull/43780

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 1 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/kokkos/package.py]: If it is not required, the directive will be wrong as soon as 3.29 gets merged. Do we know why CMake 3.29 seems to work whereas CMake 3.28 does not?

Comment (johnwparent) [reply var/spack/repos/builtin/packages/kokkos/package.py]: I'm not 100% but likely https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9243 this is why 3.29 is fine and 3.28 is not. Updated the conflict.

**Review by alalazo — APPROVED** [review]

**Pull request merged**